File: Python/02_Linked_List/LinkedListT.py
import LinkedListModule

class LinkedListT(LinkedListModule.LinkedList):

    def append(self, data):
        if self.head is None:
            self.head = LinkedListModule.Node(data)
            self.tail = self.head
            self.length += 1
            return
        temp = LinkedListModule.Node(data)
        self.tail.next = temp
        self.tail = self.tail.next
        self.length+=1
        return


    def setTail(self, node):
        self.tail.next = node
        self.tail = self.tail.next

    def __init__(self, list_length=None, circular=False, k=None, init_vals=None):
        if list_length is not None:
            self.head = None
            self.length = 0
            self.tail = None
            if init_vals is not None:
                for i in range(0, list_length):
                    self.append(init_vals[i])
            else:
                for i in range(0, list_length):
                    self.append(i)
                if circular:
                    if k is None:
                        self.tail.next = self.head
                    if k is not None:
                        k = k % list_length # in case the user enters k = 20 on a list that's 10 long
                        n = self.head
                        count = 0
                        while(n != None and k > count):
                            n = n.next
                            count += 1
                        self.tail.next = n
        else:
            LinkedListModule.LinkedList.__init__(self)
            self.tail = None
            self.length = 0
        return


    def prepend(self, data):
        if self.head == None:
            self.head = LinkedListModule.Node(data)
            self.tail = self.head
            self.length += 1
        else:
            temp = LinkedListModule.Node(data)
            temp.next = self.head
            self.head = temp
            self.length += 1
        return


    def reverse(self):
        retval = LinkedListT()
        temp = self.head;
        while (temp is not None):
            retval.prepend(temp.data)
            temp = temp.next
        return retval



def sum_lists(LLT1, LLT2):
    n1 = LLT1.head
    n2 = LLT2.head
    previous_quotient = 0
    retval = LinkedListT()

    while (n1 is not None or n2 is not None):
        val1 = 0
        val2 = 0
        if n1 is not None:
            val1 = n1.data
        if n2 is not None:
            val2 = n2.data
        the_sum = val1 + val2 + previous_quotient
        remainder = the_sum % 10
        retval.append(remainder)
        previous_quotient = the_sum // 10
        
        if n1 is not None:
            n1 = n1.next
        if n2 is not None:
            n2 = n2.next

    if previous_quotient > 0:
        retval.append(previous_quotient)

    return retval

File: C++/Practice_Pad/source/CH2/LinkedListT.cpp
#include "pch.h"
#include "LinkedListT.h"
#include <iostream>

LinkedListT::LinkedListT() {
    LinkedList();
}

LinkedListT::LinkedListT(int* init_vals, int len) {

    Node* temp = head;
    for (int i = 0; i < len; i++) {
        append(init_vals[i]);
    }
}

void LinkedListT::copy(const LinkedListT& other) {
    
    Node* temp = other.head;
    while (temp != NULL) {
        append(temp->data);
        temp = temp->next;
    }
}

LinkedListT::LinkedListT(const LinkedListT& other) {
    copy(other);
}

LinkedListT& LinkedListT::operator=(const LinkedListT& other) {

    if (&other != this) {
        cleanup();
        copy(other);
    }
    return *this;
}

LinkedListT::LinkedListT(int len, bool circular) {

    for (int i = 0; i < len; i++) {
        append(i);
    }
    if (circular) { //for testing loop function only
        tail->next = head;
        bad_list = true;
    }
}

// only used to create intersecting LL
void LinkedListT::setTailNext(Node* node) {
    tail->next = node;
    tail = tail->next;
    bad_list = true;
}

void LinkedListT::append(int data) {

    if (head == NULL) {
        head = new Node(data);
        tail = head;
        length++;
        return;
    }
    Node* temp = new Node(data);
    tail->next = temp;
    tail = temp;
    length++;
    return;
}

void LinkedListT::prepend(int data) {
    if (head == NULL) {
        head = new Node(data);
        tail = head;
    } else {
        Node* temp = new Node(data);
        temp->next = head;
        head = temp;
    }
    length++;
}

LinkedListT LinkedListT::reverse() {
    Node* temp = head;
    LinkedListT retval = LinkedListT();
    while (temp != NULL) {
        retval.prepend(temp->data);
        temp = temp->next;
    }
    return retval;
}

void LinkedListT::cleanup() {
    if (head == nullptr) {
        return;
    }
    if (bad_list) {

        Node* current = head;
        Node* next;
        for (int i = 0; i < length; i++) {

            if (current != NULL) {
                next = current->next;
                delete current;
                current = next;
            }
            else {
                break;
            }
        }
        tail = NULL;
        head = NULL;
        length = 0;
        return;
    }

    Node* current = head;
    Node* next;
    while (current != NULL) {
        next = current->next;
        delete current;
        current = next;
    }
    length = 0;
    head = NULL;
    tail = NULL;
}

LinkedListT::~LinkedListT() {
    cleanup();
}
LinkedListT LinkedListT::sum_lists(const LinkedListT& ll) const {

    LinkedListT retval = LinkedListT();
    Node * n1 = head;
    Node * n2 = ll.head;

    int remainder = 0;
    int prev_quotient = 0;

    while ((n1 != NULL) || (n2 != NULL)) {
        int val1 = 0;
        int val2 = 0;
        if (n1 != NULL){
            val1 = n1->data;
        }
        if (n2 != NULL) {
            val2 = n2->data;
        }
        int sum = val1 + val2 + prev_quotient;
        remainder = sum % 10;
        
        retval.append(remainder);

        prev_quotient = sum / 10;
        if (n1 != NULL){
            n1 = n1->next;
        }
        if (n2 != NULL) {
            n2 = n2->next;
        }
    }
    if (prev_quotient > 0) {
        retval.append(prev_quotient);
    } 
    return retval;
}

